### DISCUSSION TITLE
Change the default material on empty material slots to exporter aborting

### DISCUSSION THREAD

**Discussion by ssnd292** [StjerneIdioten/I3D-Blender-Addon#115] [Ideas & Feature Requests]
Regarding https://github.com/StjerneIdioten/I3D-Blender-Addon/pull/114

I've always had the standpoint of educating users by forcing them to correctly prepare models for exporting by not allowing bad practices.
In Blender, one of them is having a model with empty material slots. While the idea behind those in Blender themselves are questionable, it's an easy fix itself.
Putting another material into those empty slots does not fix the problem - it convolutes the object data and adds more drawcalls than necessary.

So instead of simply assigning a default material to an object that has an empty material slot, the exporter should:
...stop exporting at all and give a warning in the log and display
or
... stop exporting that particular object with that empty slot.

(Don't confuse this with having a mesh with no material slot at all - adding a default material here is just fine for debug or prototype purposes)

**Comment by kbrandwijk** [comment]
I support the idea, but I do think a change in architecture to 'validate' the scene before starting the export would probably be needed (or some kind of dry-run like implementation), because I don't think the current structure supports breaking off the process like that very well, especially not that deep in the call hierarchy.

**Reply by StjerneIdioten** [reply]
Which is also the reason for my closing statement in #114.

The current architecture has the idea of "every object is handled once", which means that validation checks happen as it goes through the objects and xml is generated as part of object instantiation. This was originally implemented as a performance thing, to avoid looping over objects a ton of times.
The downside is that this makes it a bigger cleanup task to abort the export of an object mid object, especially if you are deep into it and actually exporting another object as part of the process (eg. Shapes)
Another downside is that objects can't be reused for import, which is a future goal.

This stuff will be fixed with an architecture rewrite, that I am currently planning out, which will fix a number of problems with both unit testing, future features and aborting mid export.
And hopefully also documenting the architecture this time around 😄
